docs/rfcs/029-getpage-throttling.md

```diff
@@ -0,0 +1,181 @@
+Neon Autoscaling will make this easy, [eventually](https://github.com/neondatabase/neon/pull/3913).
+
+## The Design Space
+
+What that remains is the question about *policy* and *mechanism*:
+
+**Policy** concerns itself with the question of what limit applies to a given connection|timeline|tenant.
+Candidates are:
+
+* hard limit, same limit value per connection|timeline|tenant
+    * Per-tenant will provide an upper bound for the impact of a tenant on a given Pageserver instance.
+      This is a major operational pain point / risk right now.
+* hard limit, configurable per connection|timeline|tenant
+    * This outsources policy to console/control plane, with obvious advantages for flexible structuring of what service we offer to customers.
+    * Note that this is not a mechanism to guarantee a minium provisioned rate, i.e., this is not a mechanism to guarantee a certain QoS for a tenant.
+* fair share among active connections|timelines|tenants per instance
+    * example: divide available IOPS evenly among the tenants that were active in the last 5min
+    * needs definition of "active", and knowledge of available IOPS in advance
+* ...
+
+
+Regarding **mechanism**, it's clear that **backpressure** is the way to go.
+However, we must choose between
+* **implicit** backpressure through pq/TCP and
+* **explicit** rejection of requests + retries with exponential backoff
+
+Further, there is the question of how throttling GetPage@LSN will affect the **internal GetPage latency SLO**:
+where do we measure the SLI for Pageserver's internal getpage latency SLO? Before or after the throttling?
+
+And when we eventually move the measurement point into the Computes (to avoid coordinated omission),
+how do we avoid counting throttling-induced latency toward the internal getpage latency SLI/SLO?
+
+## Scope Of This RFC
+
+**This RFC proposes introducing a hard GetPage@LSN/second limit per tenant, with the same value applying to each tenant on a Pageserver**.
+
+This proposal is easy to implement and significantly de-risks operating large Pageservers,
+based on the assumption that extremely-high-GetPage-rate-episodes like the one from the "Motivation" section are uncorrelated between tenants.
+
+For example, suppose we pick a limit that allows up to 10 tenants to go at limit rate.
+Suppose our Pageserver can serve 100k GetPage/second total at a 100% page cache miss rate.
+If each tenant gets a hard limit of 10k GetPage/second, we can serve up to 10 tenants at limit speed without latency degradation.
+
+The mechanism for backpressure will be TCP-based implicit backpressure.
+The compute team isn't concerned about prefetch queue depth.
+Pageserver will implement it by delaying the reading of requests from the libpq connection(s).
+Pageserver will implement starvation prevention, but not guaranteed fairness, among connections by using a fair semaphore.
```

Comment:
This fair semaphore would be the basis of some token bucket impl?  If you have a particular library/approach in mind for the actual rate limiting logic it would be worth mentioning here.

Would be worth thinking about how we can mitigate the performance impact of any shared structure used for rate limiting: perhaps by individual connections buffering some tokens to avoid hitting an atomic on every request.

Reply:
This RFC proposes per-tenant limiting, so, the rate limiting state would be per-tenant.
Any perf impact at a high amount of connections per tenant would be limited to the tenant.
I'm not too worried about this.
Are you?

It would be a different story if we were to introduce a global limit.

I'm thinking of using https://docs.rs/leaky-bucket/1.0.1/leaky_bucket/ per tenant.
A quick peek into the source code shows it's using parking_lot mutex for internal synchronization.
So yeah, this would likely not be suitable for a globla limit.
But for per-tenant, I think it'll be fine.

Reply:
> Are you?

A bit, but not to the point of blocking this work.  Postgres does usually open multiple connections to the pageserver when it has many clients of its own as I understand it.  The cost is likely to be comparable to if we were doing extra Arc::clone of the tenant on each request: I would prefer to avoid it, but it's not going to cause a huge issue.

I do prefer to avoid this kind of thing pre-emptively: once we add this kind of cost, it is very hard to later measure it, because the main impact of heavy non-mutex atomics is just general "cache pain" rather than some specific profile-able issue.  But that doesn't have to happen right now: we can always later add a per-connection-handler token cache if we want to.

Reply:
>  Postgres does usually open multiple connections to the pageserver when it has many clients of its own as I understand it.

Even a single query can result in many open connections, due to the postgres `parallel_workers_*` ...

>  do prefer to avoid this kind of thing pre-emptively: once we add this kind of cost, it is very hard to later measure it, because the main impact of heavy non-mutex atomics is just general "cache pain" rather than some specific profile-able issue. 

Yep.

I figure at some point we'll need to use VTune to understand where the CPU cycles go.

But, keep in mind we already have (global!) atomics on the hotpath, namely, the prometheus metrics. (I wish they were per-cpu counters)

Reply:
https://github.com/neondatabase/neon/pull/5648/commits/02ce2d338707f61bc84b840d036c10b4c538ae9e